src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,245 @@
+card_subtype_mono _ _ $ λ x H1 H2, H1 $ congr_arg (f x) H2
+
+lemma hamm_dist_comp (f : Π i, γ i → β i) {x y : Π i, γ i} (hf : Π i, injective (f i))
+  : hamm_dist (λ i, f i (x i)) (λ i, f i (y i)) = hamm_dist x y :=
+le_antisymm (hamm_dist_comp_le_hamm_dist  _) (card_subtype_mono _ _ (λ i H1 H2, H1 (hf i H2)))
+
+lemma hamm_dist_smul_le_hamm_dist [Π i, has_scalar α (β i)] {k : α} {x y : Π i, β i} :
+  hamm_dist (k • x) (k • y) ≤ hamm_dist x y :=
+hamm_dist_comp_le_hamm_dist $ λ i, (•) k
+
+lemma hamm_dist_smul [Π i, has_scalar α (β i)] {k : α} {x y : Π i, β i}
+  (hk : Π i, is_smul_regular (β i) k) : hamm_dist (k • x) (k • y) = hamm_dist x y :=
+hamm_dist_comp (λ i (c : β i), k • c) hk
+
+@[simp] lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
+by {  rw [function.funext_iff, hamm_dist, card_eq_zero_iff],
+      exact ⟨ λ h i, imp_of_not_imp_not _ _ (λ H, h.elim' ⟨i, H⟩) h,
+              λ h, subtype.is_empty_of_false (λ i H, H (h _))⟩ }
+
+@[simp] lemma hamm_dist_self (x : Π i, β i) : hamm_dist x x = 0 := hamm_dist_eq_zero.mpr rfl
+
+lemma hamm_dist_comm (x y : Π i, β i) : hamm_dist x y = hamm_dist y x :=
+by simp_rw [hamm_dist, ne_comm]
+
+lemma hamm_dist_triangle (x y z : Π i, β i) :
+  hamm_dist x z ≤ hamm_dist x y + hamm_dist y z :=
+begin
+  simp_rw hamm_dist,
+  refine le_trans (card_subtype_mono _ _ (λ _ h, _)) (card_subtype_or _ _),
+  by_contra' H, exact h (eq.trans H.1 H.2)
+end
+
+lemma eq_of_hamm_dist_eq_zero (x y : Π i, β i) : hamm_dist x y = 0 → x = y :=
+hamm_dist_eq_zero.mp
+
+lemma hamm_dist_ne_zero {x y : Π i, β i} : hamm_dist x y ≠ 0 ↔ x ≠ y :=
+not_iff_not.mpr hamm_dist_eq_zero
+
+lemma hamm_dist_pos {x y : Π i, β i} : 0 < hamm_dist x y ↔ x ≠ y :=
+by rw [←hamm_dist_ne_zero, iff_not_comm, not_lt, nat.le_zero_iff]
+
+@[simp] lemma hamm_dist_lt_one {x y : Π i, β i} : hamm_dist x y < 1 ↔ x = y :=
+by {rw nat.lt_one_iff, exact hamm_dist_eq_zero}
+
+lemma hamm_dist_eq_zero_iff_forall_eq {x y : Π i, β i} :
+  hamm_dist x y = 0 ↔ ∀ i, x i = y i :=
+by rw [hamm_dist_eq_zero, function.funext_iff]
+
+lemma hamm_dist_ne_zero_iff_exists_ne {x y : Π i, β i} :
+  hamm_dist x y ≠ 0 ↔ ∃ i, x i ≠ y i :=
+by rw [hamm_dist_ne_zero, function.ne_iff]
+
+section has_zero
+
+variables [Π i, has_zero (β i)] [Π i, has_zero (γ i)]
+
+/-- The Hamming weight function to the naturals. -/
+
+def hamm_wt (x : Π i, β i) : ℕ := hamm_dist x 0
+
+lemma hamm_wt_eq (x : Π i, β i) : hamm_wt x = card {i // x i ≠ 0} := rfl
+
+lemma hamm_wt_eq_hamm_dist_zero (x : Π i, β i) : hamm_wt x = hamm_dist x 0 := rfl
+
+lemma hamm_wt_nonneg {x : Π i, β i} : 0 ≤ hamm_wt x := hamm_dist_nonneg
+
+lemma hamm_wt_le_card_fintype {x : Π i, β i} : hamm_wt x ≤ card ι := hamm_dist_le_card_fintype
+
+lemma hamm_wt_comp_le_hamm_wt (f : Π i, γ i → β i) {x : Π i, γ i} (hf : Π i, f i 0 = 0) :
+  hamm_wt (λ i, f i (x i)) ≤ hamm_wt x :=
+by {  simp_rw hamm_wt_eq_hamm_dist_zero, convert hamm_dist_comp_le_hamm_dist f,
+      simp_rw [pi.zero_apply, hf], refl }
```

Comment:
Again, this style isn't idiomatic; using `begin`/`end` is preferred. You can also golf slightly here to
```suggestion
begin
  refine eq.trans_le _ (hamm_dist_comp_le_hamm_dist f),
  simp_rw [hamm_wt, pi.zero_def, hf],
end
```

Reply:
Oh, my bad.